nexus/mgs-updates/src/test_util/step_through.rs

```diff
@@ -0,0 +1,144 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use futures::future::BoxFuture;
+use std::sync::Arc;
+use std::task::{Context, Poll, Wake, Waker};
+use tokio::sync::mpsc;
+
+/// Steps through *wakeups* of a `Future`
+pub struct StepThrough<'a, T> {
+    fut: BoxFuture<'a, T>,
+}
+
+impl<'a, T> StepThrough<'a, T> {
+    pub fn new(fut: BoxFuture<'a, T>) -> StepThrough<'a, T> {
+        StepThrough { fut }
+    }
+
+    /// Run the future to the next await point where it blocks, then also wait
+    /// for it to be ready to run again.
+    ///
+    /// This is not quite the same as running to an await point, since it's
+    /// possible that the future doesn't block at the await point.  In that
+    /// case, it will keep running and this function will not notice until the
+    /// next time the function does block.
+    pub async fn step(mut self) -> StepResult<'a, T> {
+        let (tx, mut rx) = mpsc::channel(16);
+        let my_waker = StepThroughWaker(tx);
+        let waker = Waker::from(Arc::new(my_waker));
+        let mut context = Context::from_waker(&waker);
+
+        // This is an oversimplification, but you can think of this like:
+        //
+        // - On entry to this function, the future is generally ready to run (to
+        //   its next await point).
+        // - This function polls the future, causing it to run until it either
+        //   completes or reaches an await point.
+        // - If the future reaches an await point, this function waits for the
+        //   future to be woken up again.
+        // - Then this function returns, preserving the invariant that on entry,
+        //   it's ready to run again.
+        match self.fut.as_mut().poll(&mut context) {
+            Poll::Pending => {
+                // The future has reached an await point.  We want to wake up
+                // when it gets woken up.  We gave its `poll()` a waker that
+                // will send a message on this channel when that happens.
+                //
+                // This only fails if the tx side has been dropped, which means
+                // the waker was dropped.  It's not clear how this could happen
+                // unless the future was being dropped, in which case this
+                // object must also be dropped.
```

Comment:
If this reasoning is wrong I think we'd go into a busy loop constantly polling the future, right? I don't think the reasoning is wrong, though - maybe we should panic on an error from this recv()?